### PR TITLE
Custom CSV Generation logic

### PR DESCRIPTION
**Pull request by davidvossel**
This tool allows us to inject custom information related to the package into the CSV, like description, keywords, maintainers, replaces version, etc...

This logic works by generating a temporary CSV using the operator-sdk which the tool then reads in and injects all our custom data into. 

## Building Custom CSV

A custom CSV entry for the registry container can be generated using the
`make generate-csv` make target.

First export your CSV details as environment variables.

__required options__
export IMAGE_REGISTRY="quay.io"
export REGISTRY_NAMESPACE="some-operator=repo-namespace"
export IMAGE_TAG="some-operator-image-tag"
export CSV_VERSION="0.0.3"

__optional options__
export REPLACES_CSV_VERSION="0.0.2"
export CSV_SKIP_RANGE=">=0.0.1 <0.0.2"

Then run `make generate-csv`

The result will be stored in the deploy/olm-catalog/performance-addon-operator
directory within a directory that matches the CSV_VERSION set.

Running `make registry-container` after creating a new custom CSV will result
in a registry bundle that includes the new CSV version and all other CSV
versions in the deploy/olm-catalog directory.

### CODE REVIEW

**Comment by davidvossel** [comment]
/hold

I want to write a readme on this before we consider merging so it's obvious how to use it for custom builds.

**Comment by slintes** [comment]
I'm wondering how often something like this was invented already :/ Should be possible with operator-sdk out of the box.
Will do a detailed review on Monday with a fresh mind.

**Comment by davidvossel** [comment]
/hold cancel

I added the README logic. This should be ready for review now.

@fromanirh you might want to take a look at how the CSV generation works and make sure it fits the workflows you'll be working with.

File: deploy/olm-catalog/performance-addon-operator/0.0.1/performance-addon-operator.v0.0.1.clusterserviceversion.yaml [hunk 1]
```diff
@@ -42,11 +43,13 @@ spec:
   customresourcedefinitions:
     owned:
     - description: PerformanceProfile is the Schema for the performanceprofiles API.
+      displayName: Performance Profile
       kind: PerformanceProfile
       name: performanceprofiles.performance.openshift.io
       version: v1alpha1
-      displayName: placeholder
-  description: Placeholder description
+  description: |2-
```
Comment (cynepco3hahue) [inline deploy/olm-catalog/performance-addon-operator/0.0.1/performance-addon-operator.v0.0.1.clusterserviceversion.yaml]: `|2-`?

Comment (slintes) [reply deploy/olm-catalog/performance-addon-operator/0.0.1/performance-addon-operator.v0.0.1.clusterserviceversion.yaml]: | = keep new lines
2 = indent with two spaces
- = no new line at the end

yaml is fun ;)
https://yaml-multiline.info/

**Comment by fromanirh** [comment]
@davidvossel ack, will review ASAP

File: hack/csv-generate.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+
+set -e
+
+export GOROOT=$(go env GOROOT)
+TMP_CSV_VERSION="9.9.9"
```
Comment (fromanirh) [inline hack/csv-generate.sh]: why this and not, say, `0.0.0`? maybe you fixed already on later patches

**Comment by fromanirh** [comment]
giving other reviewers time to chime in

File: hack/csv-generate.sh (hunk 2 above)
Comment (davidvossel) [reply hack/csv-generate.sh]: it's a temporary csv that gets deleted. i can try 0.0.0 and see if that works too

File: tools/csv-generator/csv-generator.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+	err := json.Unmarshal(csv.Spec.InstallStrategy.StrategySpecRaw, templateStrategySpec)
+	if err != nil {
+		panic(err)
+	}
+
+	return templateStrategySpec
+}
+
+func marshallObject(obj interface{}, writer io.Writer) error {
+	jsonBytes, err := json.Marshal(obj)
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply tools/csv-generator/csv-generator.go]: yeah, no joke :(

**Pull request merged**